Utilities/O2Device/include/O2Device/Utilities.h

```diff
@@ -39,23 +40,35 @@ namespace Base
 using O2Message = FairMQParts;
 
 //__________________________________________________________________________________________________
-// AddDataBlock for generic (compatible) containers, that is contiguous containers using the pmr allocator
+// addDataBlock for generic (compatible) containers, that is contiguous containers using the pmr allocator
 template <typename ContainerT, typename std::enable_if<!std::is_same<ContainerT, FairMQMessagePtr>::value, int>::type = 0>
-bool AddDataBlock(O2Message& parts, o2::header::Stack&& inputStack, ContainerT&& inputData, o2::memoryResources::FairMQMemoryResource* targetResource = nullptr)
+bool addDataBlock(O2Message& parts, o2::header::Stack&& inputStack, ContainerT&& inputData, o2::memoryResources::FairMQMemoryResource* targetResource = nullptr)
 {
   using std::move;
-  auto dataMessage = getMessage(move(inputData), targetResource);
-  return AddDataBlock(parts, move(inputStack), move(dataMessage), targetResource);
+  using std::forward;
+
+  auto headerMessage = getMessage(move(inputStack), targetResource);
+  auto dataMessage = getMessage(forward<ContainerT>(inputData), targetResource);
+
+  parts.AddPart(move(headerMessage));
+  parts.AddPart(move(dataMessage));
+
   return true;
 }
 
 //__________________________________________________________________________________________________
-// AddDataBlock for data already wrapped in FairMQMessagePtr
+// addDataBlock for data already wrapped in FairMQMessagePtr
 // note: since we cannot partially specialize function templates, use SFINAE here instead
 template <typename ContainerT, typename std::enable_if<std::is_same<ContainerT, FairMQMessagePtr>::value, int>::type = 0>
-bool AddDataBlock(O2Message& parts, o2::header::Stack&& inputStack, ContainerT&& dataMessage, o2::memoryResources::FairMQMemoryResource* targetResource = nullptr)
+bool addDataBlock(O2Message& parts, o2::header::Stack&& inputStack, ContainerT&& dataMessage, o2::memoryResources::FairMQMemoryResource* targetResource = nullptr)
 {
   using std::move;
+
+  //make sure the payload size in DataHeader corresponds to message size
+  using o2::header::DataHeader;
+  DataHeader* dataHeader = const_cast<DataHeader*>(o2::header::get<DataHeader*>(inputStack.data()));
```

Comment:
we have to change `o2::header::get` to avoid the `const_cast`. It's legal to change `payloadSize` because `inputStack` is passed by move and thus owned by the function. should add const header only if the buffer it is queried from is const.

Reply:
@matthiasrichter agreed. This PR or a different one? I was also thinking to add a default value for the size in the DataHeader ctor so it can be omitted since it is always(?) possible to set the correct size automatically, which is safer

Reply:
no, this was more a general comment, would do this in a separate PR